File: menu-service/src/main/java/com/acme/menusrv/dish/service/DishService.java
package com.acme.menusrv.dish.service;

import com.acme.commons.security.NotAccountantAuthenticated;
import com.acme.menusrv.dish.dto.CreateDishDto;
import com.acme.menusrv.dish.dto.DishDto;
import com.acme.menusrv.dish.dto.DishFilter;
import com.acme.menusrv.dish.dto.DishNameDto;
import com.acme.menusrv.dish.dto.FullDetailsDishDto;
import com.acme.menusrv.dish.dto.UpdateDishDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.validation.annotation.Validated;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import javax.validation.Valid;
import java.util.List;
import java.util.UUID;

@Validated
public interface DishService {
    @PreAuthorize("hasAnyAuthority('ADMIN','COMPANY_OWNER','PP_MANAGER','CHEF')")
    Mono<UUID> create(@Valid CreateDishDto createDto);

    @PreAuthorize("hasAnyAuthority('ADMIN','COMPANY_OWNER','PP_MANAGER','CHEF')")
    Mono<Void> update(UUID id, @Valid UpdateDishDto updateDto);

    @PreAuthorize("hasAnyAuthority('ADMIN','COMPANY_OWNER','PP_MANAGER','CHEF')")
    Mono<Void> delete(UUID id);

    @NotAccountantAuthenticated
    Mono<DishDto> findById(UUID id);

    @PreAuthorize("hasAnyAuthority('ADMIN','COMPANY_OWNER','PP_MANAGER','CHEF','COOK')")
    Mono<FullDetailsDishDto> findFullDetailsById(UUID id);

    @PreAuthorize("hasAnyAuthority('ADMIN','COMPANY_OWNER','PP_MANAGER','CHEF','COOK')")
    Mono<Page<DishDto>> find(@Valid DishFilter filter, Pageable pageable);

    @PreAuthorize("hasAnyAuthority('ADMIN','COMPANY_OWNER','PP_MANAGER','CHEF','COOK')")
    Mono<List<String>> findTags(UUID companyId, UUID publicPointId);

    @PreAuthorize("hasAnyAuthority('ADMIN','COMPANY_OWNER','PP_MANAGER','CHEF','COOK')")
    Flux<DishNameDto> findNames(UUID companyId, UUID publicPointId);
}

File: admin-ui/src/dish/dishService.js
import restApi, { buildGetFilterParams } from '../common/restApi'

const BASE_URL = '/menu-service/api/dishes'

const dishService = {
    find: function (filter, pageable, sort) {
        return restApi.get(BASE_URL, {
            params: buildGetFilterParams(filter, pageable, sort)
        })
    },
    findById: function (id) {
        return restApi.get(`${BASE_URL}/${id}`)
    },
    findFullDetailsById: function (id) {
        return restApi.get(`${BASE_URL}/${id}/full-details`)
    },
    update: function (id, request) {
        return restApi.put(`${BASE_URL}/${id}`, request)
    },
    create: function (request) {
        return restApi.post(BASE_URL, request)
    },
    delete: function (id) {
        return restApi.delete(`${BASE_URL}/${id}`)
    },
    findTags(companyId, publicPointId) {
        return restApi.get(`${BASE_URL}/tags`, {
            params: { companyId, publicPointId }
        })
    },
    findNames(companyId, publicPointId) {
        return restApi.get(`${BASE_URL}/names`, {
            params: { companyId, publicPointId }
        })
    }
}

export default dishService